services/gundeck/test/unit/MockGundeck.hs

```diff
@@ -0,0 +1,719 @@
+      . fmap removeSelf
+      . fmap insertAllClients
+      $ rcps
+      where
+        reachable :: ((UserId, ClientId), payload) -> Bool
+        reachable (ids, _) = wsReachable env ids
+
+        removeSelf :: ((UserId, Maybe ClientId, any), (Recipient, Payload)) -> (Recipient, Payload)
+        removeSelf ((_, Nothing, _), same) =
+          same
+        removeSelf ((_, Just sndcid, _), (Recipient rcpuid route cids, pay)) =
+          (Recipient rcpuid route $ filter (/= sndcid) cids, pay)
+
+        insertAllClients :: (any, (Recipient, Payload))
+                         -> (any, (Recipient, Payload))
+        -- if the recipient client list is empty, fill in all devices of that user
+        insertAllClients (same, (Recipient uid route [], pay)) = (same, (rcp', pay))
+          where rcp' = Recipient uid route (clientIdsOfUser env uid)
+
+        -- otherwise, no special hidden meaning.
+        insertAllClients same@(_, (Recipient _ _ (_:_), _)) = same
+
+    expectNative :: MockEnv -> NotifQueue
+    expectNative env
+      = foldl' (uncurry . deliver) mempty
+      . filter reachable
+      . reformat
+      . mconcat . fmap removeSome
+      . mconcat . fmap insertAllClients
+      $ rcps
+      where
+        reachable :: ((UserId, ClientId), payload) -> Bool
+        reachable (ids, _) = nativeReachable env ids && not (wsReachable env ids)
+
+        removeSome :: ((UserId, Maybe ClientId, Bool), (Recipient, Payload)) -> [(Recipient, Payload)]
+        removeSome ((_, _, _), (Recipient _ RouteDirect _, _)) =
+          -- never native-push recipients marked 'RouteDirect'.
+          []
+        removeSome ((snduid, _, False), same@(Recipient rcpuid _ _, _)) =
+          -- if pushNativeIncludeOrigin is False, none of the originator's devices will receive
+          -- native pushes.
+          [same | snduid /= rcpuid]
+        removeSome ((_, Just sndcid, True), (Recipient rcpuid route cids, pay)) =
+          -- if originating client is known and pushNativeIncludeOrigin is True, filter out just
+          -- that client, and native-push to all other devices of the originator.
+          [(Recipient rcpuid route $ filter (/= sndcid) cids, pay)]
+        removeSome ((_, Nothing, True), same) =
+          -- if NO originating client is known and pushNativeIncludeOrigin is True, push to all
+          -- originating devices.
+          [same]
+
+        insertAllClients :: (any, (Recipient, Payload))
+                         -> [(any, (Recipient, Payload))]
+        -- if the recipient client list is empty, fill in all devices of that user
+        insertAllClients (same, (Recipient uid route [], pay)) = [(same, (rcp', pay))]
+          where rcp' = Recipient uid route (clientIdsOfUser env uid)
+        -- otherwise, no special hidden meaning.
+        insertAllClients same@(_, (Recipient _ _ (_:_), _)) = [same]
+
+    expectCass :: NotifQueue
+    expectCass = foldl' (uncurry . deliver) mempty . mconcat $ go <$> pushes
```

Comment:
Is there a reason why `insertAllClients` is not used here? (If it's an omission, I'll fix it.)

Reply:
i think it may be an oversight.  (i wonder if we need to improve coverage by the Arbitrary generators further to catch this?)

btw i think there is some more code duplication to get rid of, but you're probably on it.